README.md

TECOC - TECO Editor
===================

Teco is a text editor written by Dan Murphy in 1962.  It differs from
other editors in several ways as follows:

### Character oriented

Vi and Emacs are screen oriented editors.  You view the text on the whole
window or screen, and you edit the text right where you see it.  Changes take
affect immediately.

ed, ex, edlin are line oriented editors.  Edited text is viewed,
referenced, and edited by line.

Teco is character oriented.  This means the unit of reference is a
single character.  On one hand, this may make teco cumbersome to use,
but on the other hand, it gives teco very fine control over what is
happening.

Teco does have a video mode, however text is still edited with
commands on a character basis.  The screen is simply updated once a
command is completed to give the user a visual representation of where
they are.  Text is not edited in the screen view area.

### Macros

In general, vi, ed, and edlin are not programmable.  This means you
have the functions the editor provides and that is all you can do.
However, teco is highly programmable.  Programs that you write are
called macros.  In fact, the original version of Emacs was written as
teco macros.  Also, "emacs" actually means "Editing Macros".

### TECOC

The original versions of Teco were written in the machine languages of
the machines they operated on.  As newer machines and editors arose, teco
fell into increasing disuse and obscurity.

In the early '80's Pete Siemsen, a long-time teco user, created a
portable version of Teco in the C language called "tecoc".  Rather
than create a new editor, Pete's goal was to duplicate the original teco
as closely as possible.

As time wore on, Tom Almy, Blake McBride, and others increasingly
assisted in tecoc's maintenance (porting, bug fixes, etc.).
Eventually, tecoc found a new home and principal support at Tom
Almy's home page where Tom kept up with ports to Windows, Mac, and
Linux.

Most recently, Blake McBride:

1. Merged Tom Almy's various ports along with work done by Blake
2. renewed the Windows, Mac, and Linux ports
3. re-enabled video support (on Mac & Linux)
4. bug fixes and speed improvements
5. tested on 64 bit machines

This latest work is available on github at:

https://github.com/blakemcbride/TECOC

Other documents in the "doc" directory provide complete documentation.
See the following files in the "doc" directory:

File | Description
:------- | :----------------
BUILD.txt | Build and setup instructions
INTRO.txt | Introduction to using tecoc
teco-manual.txt | Complete manual


doc/changes.txt
	This file contains release notes for TECOC.  The "base version"
is 100 (reported by the EO command).  The version number starts at 100 to
avoid confusion with TECO-11 version numbers,  which were close to 40 as of
01-Jan-1987.


10-Dec-1987 version 100

baseline version as of Fall Decus Symposium, Anaheim


04-Jan-1988 version 101		(mostly stuff for XENIX)

1. created this file
2. changed EXEEO.C to make the version number 101.
3. changed FBFSIZ (file name size) to 20 in AMODEM.C
4. changed XAFTER.CSH to execute under CSH instead of SH
5. removed unused tags IFTag and OFTag from ZPORT.H
6. added comment about this file to AAAREADME.TXT
7. Created an updated XBUILD.CSH.
8. added a tab to INIT.C.
9. Added debugging code to PARSCL.C, MEMINI.C, ZINIFL.C, MUNG.C, ZREMBR.C,
   and INIOPN.C.  This code is activated only if the variable DbgLvl is
   initialized to 1 in DBGDSP.C.
10. Added semi-colon to line 181 of ZOPINP.C so it'll compile.
11. Changed ZREMBR.C so it returns NO in XENIX section.
12. Added code to set TrmTyp in XENIX section of ZTRMNL.C.
13. Changed SCRNOP.C so it tests TrmTyp (not OpCode) against UNTERM.
14. Added XENIX sections to ZDSPCH.C, ZDSPBF.C and ZCHIN.C.


23-Jan-1988 version 102

1. changed EXEEO.C to make the version number 102.
2. changed the debugging code in ZABORT.C to avoid an infinite loop
3. removed "signal(SIGTERM, ZAbort)" from XENIX code in ZINIT.C
4. changed ZTRMNL.C to support control-Cs under XENIX
5. check for control-C in ZCHIN.C,  needed to do it in all other Z routines
6. added XENIX section to ZRDLIN.C
7. define sys$close in ZOPINP.C and ZOPOUT.C
8. define sys$get, sys$putmsg in ZRDEI.C
9. renamed/modified all the .CSH and .COM files so they start with "Z" and
   updated AAAREADME.TXT to reflect the changes.
10. added file FALL87.RNO,  updated AAAREADME.TXT about it
11. fixed comment in RDLINE.C.
12. changed || to | in ZHELP.C so / immediate commands are okay under VMS.
13. fixed "invalid buffer size" bug for reading large files by changing
    "65536" to "65535" in ZRDLIN.C.
14. fixed control-C handling,  so that type-out of a large buffer works.
    Changed EXECST.C, ZDSPBF.C, FRSTCH.C, READCS.C.
15. added a test to EXECST.C so that commands like "<test$$" will
    now produce "unterminated command" errors,  as they should
16. modified EXEEI.C to make EI$$ a null command to make CHANGE.TEC work


30-Jan-1988 version 103		( to make CHANGE.TEC work )

1. changed EXEEO.C to make the version number 103.
2. changed definition of Mung from DEFAULT to VOID in MUNG.C.
3. changed TmpEI from DEFAULT to VOID in TMPEI.C, ZINIFL.C, MUNG.C
4. removed superfluous "else" from ZSWILD.C
5. moved WFab, WNam initialization from ZINIFL.C to ZINIT.C
6. set WFab.fab$l_fna in ZINIT.C to make EN command work
7. added XENIX section to ZCPYBL.C
8. fixed XENIX section of ZICLOS.C
9. fixed XENIX section of ZOPOUT.C
10. fixed XENIX section of ZRDLIN.C
11. fixed XENIX section of ZWRITE.C
12. changed constants for XENIX in ZPORT.H
13. changed calloc to malloc in ZALLOC.C, merged similar sections
14. added debugging code to RDLINE.C
15. added debugging line to WRBUFR.C
16. fixed unlink call in ZOCLOS.C,  added debugging code
17. fixed unlink call in ZOCLDE.C
18. added -O switch to CC commands in XMKBLD.CSH (optimizes code)


31-Mar-1988 version 104

1. changed EXEEO.C to make the version number 104.
2. changed the way code is compiled under XENIX:  deleted XMKBLD.CSH,
   created ZXBLD.CSH and ZXDBLD.CSH,  modified ZXLOGIN.CSH and ZXCCC.CSH
3. fixed XENIX file creation mode in ZOPOUT.C
4. fixed XENIX section of ZMKOFN.C
5. fixed setting of CmdMod and EStTop in EXEEB.C
6. removed unnecessary call to ExeEY in INIOPN.C
7. fixed XENIX file renaming in ZOCLOS.C
8. used auto-decrement instead of addition in MEMINI.C
9. Renamed CHRFUNCS.H to CHRMACS.H.  All references to CHRFUNCS.H
   had to be changed:  many files modified.  Also changed macros that
   are in CHRMACS.H to use an array.  Merged that array with the one that
   was in ReadCS.
10. made ZTrmnl define TrmTyp as UNTERM when we're running in a batch job.
11. took out "dying" message in GETNMA.C
12. fixed CR/LF appending in EXEEQU.C
13. fixed file output code in ZDSPCH.C


13-Apr-1988 version 105

1. changed EXEEO.C to make the version number 105.
2. removed extra tab from ZINIT.C
3. renamed XAFTER.CSH to ZXAFTER.CSH
4. changed ZXAFTER.CSH so it renames CHRMACS.H,  not CHRFUNCS.H
5. added improved error reporting in ZOPINP.C
6. added "cc -c tecoc.c" to ZXBLD.CSH
7. added "/STANDARD=PORTABLE" qualifiers to CC commands in ZVBLD.COM
8. Fixed ZWrite to append LFs correctly under XENIX
9. Fixed ZChIn to make CR/LFs under XENIX out of LFs.
10. changed ZOpOut to use fopen instead of creat/fdopen.
11. overhauled error reporting.  Many modules changed, including all "Z" files.
    Added OpnOut and ZErMsg.
12. Replaced GroEBf with InsStr to merge several copies of the same code
    sequence.  Affected ExeBSl, ExeCtI, ExeG, ExeI, Replac.
13. Removed post-copy of memory after ZRaloc calls.  ZRaloc does it.
14. Took /STANDARD=PORTABLE off VMS compile commands to prevent warnings
    about "globalref".


30-Jul-1988 version 106

1. changed EXEEO.C to make the version number 106.
2. fixed some error reporting bugs.
3. check for mung mode failure in ExeCSt, not ErrMsg
4. changed EI buffer size to 20000 to let TSTSRC.TEC be input


06-Aug-1988 version 107

1. changed EXEEO.C to make the version number 107.
2. fixed FR command.
3. fixed it so two control-C's terminate TECOC,  as it should be.
4. overhauled command-line parsing to make it use a TECO macro to do it,
   like TECO-11 does.  Major changes to initialization code.  I did this to
   provide the extra power and to make invocation of TECOC the same as TECO-11
   under VAX/VMS.  Created ExeEG,  ZExeEJ, ZEgIni, ZEgLib, ZEgMem, ZEgSym
   and ZEgVte.  Deleted ZIniFl, Mung, ZExeEG.
5. fixed ::S command.
6. fixed ] command's memory deallocation.



src/err.c
/*****************************************************************************
	ErrChr(), ErrMsg(), ErrStr(), ErrUTC(), ErrVrb()
	These functions display TECO's error messages.
	ErrChr	display a message that contains a special character
	ErrMsg	display a message given the message number
	ErrStr	display a message that contains a special string
	ErrUTC	display UTC or UTM based on macro stack depth
	ErrVrb	display the verbose form of an error message
*****************************************************************************/
#include "zport.h"		/* define portability identifiers */
#include "tecoc.h"		/* define general identifiers */
#include "defext.h"		/* define external global variables */
#include "deferr.h"		/* define error message codes */
#include "dchars.h"		/* define identifiers for characters */
#include <string.h>		/* prototype for strlen() */
#if USE_PROTOTYPES
static	VVOID	ErrDsp(WORD ErrNum, unsigned char *ELine);
static	VVOID	ToErr(unsigned char Charac, charptr *PtrPtr);
#endif
/****
   IMPORTANT: if you change this list,  you must also change DEFERR.H,
   TECOC.RNH and VRBMSG.H !!!
****/
static char *Errors[] = {
/*  0*/  "ARG   Improper arguments",
/*  1*/  "BNI   > not in iteration",
/*  2*/  "DTB   Delete too big",
/*  3*/  "FNF   File not found \"%s\"",
/*  4*/  "ICE   Illegal ^E command in search argument",
/*  5*/  "IEC   Illegal character \"%s\" after E",
/*  6*/  "IFC   Illegal character \"%s\" after F",
/*  7*/  "IIA   Illegal insert argument",
/*  8*/  "ILL   Illegal command \"%s\"",
/*  9*/  "ILN   Illegal number",
/* 10*/  "IPA   Negative or 0 argument to P",
/* 11*/  "IQC   Illegal \" character",
/* 12*/  "IQN   Illegal Q-register name \"%s\"",
/* 13*/  "IRA   Illegal radix argument to ^R",
/* 14*/  "ISA   Illegal search argument",
/* 15*/  "ISS   Illegal search string",
/* 16*/  "IUC   Illegal character \"%s\" following ^",
/* 17*/  "MAP   Missing '",
/* 18*/  "MEM   Memory overflow",
/* 19*/  "NAB   No argument before ^_ ",
/* 20*/  "NAC   No argument before ,",
/* 21*/  "NAE   No argument before =",
/* 22*/  "NAP   No argument before )",
/* 23*/  "NAQ   No argument before \"",
/* 24*/  "NAS   No argument before ;",
/* 25*/  "NAU   No argument before U",
/* 26*/  "NCA   Negative argument to ,",
/* 27*/  "NFI   No file for input",
/* 28*/  "NFO   No file for output",
/* 29*/  "NYA   Numeric argument with Y",
/* 30*/  "NYI   Not yet implemented",
/* 31*/  "OFO   Output file already open",
/* 32*/  "PDO   Push-down list overflow",
/* 33*/  "PES   Attempt to pop an empty stack",
/* 34*/  "POP   Attempt to move pointer off page with \"%s\"",
/* 35*/  "SNI   ; not in iteration",
/* 36*/  "SRH   Search failure \"%s\"",
/* 37*/  "STL   String too long",
/* 38*/  "TAG   Missing tag !%s!",
/* 39*/  "UTC   Unterminated command",
/* 40*/  "UTM   Unterminated macro",
/* 41*/  "XAB   Execution aborted",
/* 42*/  "YCA   Y command aborted",
/* 43*/  "IFE   ill-formed numeric expression",
/* 44*/  "SYS   %s",
/* 45*/  "UCD   unable to close and delete output file %s",
/* 46*/  "UCI   unable to close input file",
/* 47*/  "UCO   unable to close output file",
/* 48*/  "UFI   unable to open file %s for input",
/* 49*/  "UFO   unable to open file %s for output",
/* 50*/  "URC   unable to read character from terminal",
/* 51*/  "URE   unable to read TECO command file",
/* 52*/  "URL   unable to read line from input file",
/* 53*/  "UWL   unable to write line to output file"
};
/*****************************************************************************
	ErrUTC()
	Simply call ErrMsg with either ERR_UTC (unterminated command) or
ERR_UTM (unterminated macro) depending on whether we're in a macro or not.
*****************************************************************************/
VVOID ErrUTC()
{
	ErrMsg((MStTop < 0) ? ERR_UTC : ERR_UTM);
}
/*****************************************************************************
	ErrVrb()
	This function displays the verbose form of an error message.
The verbose form is a short paragraph explaining the error condition in
some detail.  This function is called from ErrDsp when the least significant
two bits of the EH flag are set to 3,  or from FrstCh when the user types
a / immediately after receiving a short error message.
	Verbose explanations of error messages are stored in the HELP system
so that they can be accessed using the HELP command.  Therefore this function
calls ZHelp to display the verbose messages.
*****************************************************************************/
VVOID ErrVrb(ErrNum)		/* display verbose error message */
WORD ErrNum;			/* error message number */
{
	DBGFEN(3,"ErrVrb",NULL);
	ZVrbos(ErrNum, Errors[ErrNum]);
	DBGFEX(3,DbgFNm,NULL);
}
/*****************************************************************************
	ErrDsp()
	This function displays a TECO error message on the terminal screen.
This function is passed the number of the error message and the text of the
error message.
	The EH flag controls how verbose the error message will be.  The two
least significant bits of the EH flag control how verbose the error message
is.  The third bit (mask is EH_COMMAND) causes the erroneous command string up
to and including the erroneous command to be displayed.
	The low two bits of EH have the following meanings:
	0	same as 2
	1	only the three letter code is output
	2	three letter code and one-line error message
	3	three letter code, one-lien message and paragraph
*****************************************************************************/
static VVOID ErrDsp(ErrNum, ELine)
WORD ErrNum;			/* the error number */
unsigned char *ELine;		/* error message text */
{
	WORD HlpLvl;
	ptrdiff_t lenn;
	HlpLvl = (EhFlag & 3) ? (EhFlag & 3) : 2;
	lenn = (HlpLvl < 2) ? 4 : strlen((char *)ELine);
	ZDspBf(ELine, lenn);
	if (HlpLvl == 3) {			/* if paragraph */
		ErrVrb(ErrNum);			/* display paragraph */
	}
	ZDspBf("\r\n", 2);
	if (EhFlag & EH_COMMAND) {	/* if error string display is on */
		TypESt();
	}
	LstErr = ErrNum;
}
/*****************************************************************************
	ToErr()
	Given a character and a buffer pointer,  this function puts the
displayable representation for the character into the buffer and updates
the buffer position.  For most characters,  the displayable representation
is the character itself.
*****************************************************************************/
static VVOID ToErr(Charac, PtrPtr)
unsigned char Charac;
charptr *PtrPtr;
{
	char SBuf[5];
	char *SPtr;
	size_t SLen;
	switch (Charac) {
	    case TABCHR:	SPtr = "<TAB>";	SLen = 5; break;
	    case LINEFD:	SPtr = "<LF>";	SLen = 4; break;
	    case VRTTAB:	SPtr = "<VT>";	SLen = 4; break;
	    case FORMFD:	SPtr = "<FF>";	SLen = 4; break;
	    case CRETRN:	SPtr = "<CR>";	SLen = 4; break;
	    case ESCAPE:	SPtr = "<ESC>";	SLen = 5; break;
	    case '\0':		SPtr = "<NUL>";	SLen = 5; break;
	    default:		if (Charac >= DELETE) {
					SPtr = SBuf;
					MakDBf((LONG)Charac, 16);
					SLen = (size_t) (DBfPtr - DBfBeg);
					SBuf[0] = '[';
					MEMMOVE(&SBuf[1], DBfBeg, SLen);
					SBuf[++SLen] = ']';
					++SLen;
				} else if (Charac < SPACE) {
					SPtr = SBuf;
					MEMMOVE(SBuf, "<^x>", 4);
					SBuf[2] = Charac + '@';
					SLen = 4;
				} else {
					SPtr = (char *)&Charac;
					SLen = 1;
				}
				break;
	}
	MEMMOVE(*PtrPtr, SPtr, SLen);
	*PtrPtr += SLen;
}
/*****************************************************************************
*****************************************************************************/
VVOID ErrChr(ErrNum, EChr)	/* display error message with character arg */
WORD ErrNum;			/* the error number */
unsigned char EChr;		/* character to imbed in message */
{
	char *TmpPtr;
	unsigned char ErrBuf[ERBFSIZ];
	charptr	EPtr = ErrBuf;
	*EPtr++ = '?';
	TmpPtr = Errors[ErrNum];
	while (*TmpPtr != '%') {	/* copy up to "%" */
		*EPtr++ = *TmpPtr++;
	}
	ToErr(EChr, &EPtr);		/* copy the %c character */
	TmpPtr++;			/* skip % */
	TmpPtr++;			/* skip c */
	while (*TmpPtr) {		/* copy rest of string */
		*EPtr++ = *TmpPtr++;
	}
	*EPtr = '\0';
	ErrDsp(ErrNum, ErrBuf);
}
/*****************************************************************************
*****************************************************************************/
VVOID ErrMsg(ErrNum)		/* display error message */
WORD ErrNum;			/* the error number */
{
	char *TmpPtr;
	unsigned char ErrBuf[ERBFSIZ];
	charptr EPtr = ErrBuf;
	*EPtr++ = '?';
	TmpPtr = Errors[ErrNum];
	while (*TmpPtr) {		/* copy rest of string */
		*EPtr++ = *TmpPtr++;
	}
	*EPtr = '\0';
	ErrDsp(ErrNum, ErrBuf);
}
/*****************************************************************************
*****************************************************************************/
VVOID ErrStr(ErrNum, EStr)	/* display error message with string arg */
WORD ErrNum;			/* the error number */
char *EStr;			/* string to imbed in message */
{
	char *TmpPtr;
	unsigned char ErrBuf[ERBFSIZ];
	charptr EPtr = ErrBuf;
	*EPtr++ = '?';
	TmpPtr = Errors[ErrNum];
	while (*TmpPtr != '%') {	/* copy up to "%" */
		*EPtr++ = *TmpPtr++;
	}
	while (*EStr) {			/* copy %s string */
		ToErr((unsigned char)*EStr++, &EPtr);
	}
	TmpPtr++;			/* skip % */
	TmpPtr++;			/* skip s */
	while (*TmpPtr) {		/* copy rest of string */
		*EPtr++ = *TmpPtr++;
	}
	*EPtr = '\0';
	ErrDsp(ErrNum, ErrBuf);
}
/*****************************************************************************
*****************************************************************************/
VVOID ErrPSt(ErrNum, EBeg, EEnd) /* display error message with string arg */
WORD ErrNum;			/* the error number */
charptr EBeg;			/* string to imbed in message */
charptr EEnd;			/* ptr to char after last char */
{
	char *TmpPtr;
	unsigned char ErrBuf[ERBFSIZ];
	charptr EPtr = ErrBuf;
	*EPtr++ = '?';
	TmpPtr = Errors[ErrNum];
	while (*TmpPtr != '%') {	/* copy up to "%" */
		*EPtr++ = *TmpPtr++;
	}
	while (EBeg < EEnd) {		/* copy %s string */
		ToErr(*EBeg++, &EPtr);
	}
	TmpPtr++;			/* skip % */
	TmpPtr++;			/* skip s */
	while (*TmpPtr) {		/* copy rest of string */
		*EPtr++ = *TmpPtr++;
	}
	*EPtr = '\0';
	ErrDsp(ErrNum, ErrBuf);
}
